src/goto-symex/symex_goto.cpp

```diff
@@ -316,16 +326,26 @@ void goto_symext::merge_gotos(statet &state)
   frame.goto_state_map.erase(state_map_it);
 }
 
-void goto_symext::merge_goto(const goto_statet &goto_state, statet &state)
+void goto_symext::merge_goto(goto_statet &goto_state, statet &state)
 {
   // check atomic section
   if(state.atomic_section_id != goto_state.atomic_section_id)
     throw incorrect_goto_program_exceptiont(
       "atomic sections differ across branches",
       state.source.pc->source_location);
 
-  // do SSA phi functions
-  phi_function(goto_state, state);
+  // If our destination is unreachable just skip attempting to merge names
+  // and directly move the structures in instead.
+  if(state.guard.is_false())
+  {
+    state.level2 = std::move(goto_state.level2);
+    state.propagation = std::move(goto_state.propagation);
+  }
+  else
+  {
+    // do SSA phi functions
+    phi_function(goto_state, state);
+  }
 
   merge_value_sets(goto_state, state);
```

Comment:
Apologies for late comment, just occurred to me now -- `merge_value_sets` should be moving when we have `dest.guard.is_false()` as well.

Reply:
Well, if only we had #4169 ...